Sources_v3/Utils/Internet Connection/InternetConnection.swift

```diff
@@ -0,0 +1,250 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+import Network
+
+/// A delegate to receive Internet connection events.
+protocol InternetConnectionDelegate: class {
+    /// Calls when the Internet connection status did change.
+    /// - Parameter status: an Internet connection status.
+    func internetConnectionStatusDidChange(status: InternetConnection.Status)
+    
+    /// Calls when the Internet connection monitor got an error.
+    /// For example `Reachability` didn't initiated or started notifying properly.
+    /// - Parameter error: an error.
+    func internetConnectionMonitorFailed(with error: Error)
+}
+
+/// A protocol for Internet connection monitors.
+protocol InternetConnectionMonitor {
+    /// A delegate for receiving Internet connection events.
+    var delegate: InternetConnectionDelegate? { get set }
+    
+    /// The current status of Internet connection.
+    var status: InternetConnection.Status { get }
+    
+    /// Start Internet connection monitoring.
+    func start()
+    /// Stop Internet connection monitoring.
+    func stop()
+}
+
+/// An Internet Connection monitor.
+///
+/// Basically, it's a wrapper over legacy monitor based on `Reachability` (iOS 11 only)
+/// and default monitor based on `Network`.`NWPathMonitor` (iOS 12+).
+class InternetConnection: InternetConnectionMonitor {
+    /// A shared instance of `InternetConnection`.
+    static let shared = InternetConnection()
+    
+    var delegate: InternetConnectionDelegate? {
+        get { monitor.delegate }
+        set { monitor.delegate = newValue }
+    }
+    
+    var status: InternetConnection.Status { monitor.status }
+    
+    private var monitor: InternetConnectionMonitor
+    
+    init(monitor: InternetConnectionMonitor = {
+        if #available(iOS 12, *) {
+            return Monitor()
+        } else {
+            return LegacyMonitor()
+        }
```

Comment:
You don't need to expose this in the init and keep it just an implementation detail of `InternetConnection`

Reply:
I moved inside, but leave to possibility for tests.